hudi-common/src/main/java/org/apache/hudi/common/model/HoodieAvroRecordMerger.java

```diff
@@ -43,8 +43,11 @@ public String getMergingStrategy() {
   }
 
   @Override
-  public Option<Pair<HoodieRecord, Schema>> merge(HoodieRecord older, Schema oldSchema, HoodieRecord newer, Schema newSchema, TypedProperties props) throws IOException {
-    return combineAndGetUpdateValue(older, newer, newSchema, props)
+  public Option<Pair<HoodieRecord, Schema>> merge(Option<HoodieRecord> older, Schema oldSchema, Option<HoodieRecord> newer, Schema newSchema, TypedProperties props) throws IOException {
+    if (!older.isPresent() || !newer.isPresent()) {
```

Comment:
@yihua, I understand that `keyToNewRecords` may contain extra records that are for insert only. Since `writeIncomingRecords` has inserted them, why do we need to merge logic? To support potential custom logic?

Reply:
Based on the logic of `combineAndGetUpdateValue`. When `older` record is empty, it returns Option.empty() and newer may be good; but when `older` is good, but `newer` is empty, it will throw NullPointerException. Therefore, `older` and `newer` have to be both good in order to produce valid result. Of course, the logic of `combineAndGetUpdateValue` may be wrong initially and then I can modify as you suggest.

Reply:
Looks like the logic now is more fine-grained.  Resolving this.